File: src/utils/logger.ts
/**
 * Structured logger utility for Lambda handlers
 * Provides consistent logging format with context for debugging
 */

export interface LogContext {
  requestId?: string;
  routeKey?: string;
  path?: string;
  method?: string;
  userId?: string;
  [key: string]: any;
}

export class Logger {
  private context: LogContext = {};

  /**
   * Initialize logger with request context
   */
  init(event: any, lambdaContext?: any): void {
    this.context = {
      requestId: lambdaContext?.awsRequestId || event?.requestContext?.requestId || 'unknown',
      routeKey: event?.routeKey,
      path: event?.requestContext?.http?.path || event?.rawPath,
      method: event?.requestContext?.http?.method || event?.httpMethod,
      userId: event?.requestContext?.authorizer?.jwt?.claims?.sub,
    };
  }

  /**
   * Format message with context
   */
  private formatMessage(level: string, message: string, data?: any): string {
    const logEntry = {
      timestamp: new Date().toISOString(),
      level,
      requestId: this.context.requestId,
      routeKey: this.context.routeKey,
      path: this.context.path,
      method: this.context.method,
      userId: this.context.userId,
      message,
      ...(data && { data }),
    };
    return JSON.stringify(logEntry);
  }

  /**
   * Log informational message
   */
  info(message: string, data?: any): void {
    console.log(this.formatMessage('INFO', message, data));
  }

  /**
   * Log warning message
   */
  warn(message: string, data?: any): void {
    console.warn(this.formatMessage('WARN', message, data));
  }

  /**
   * Log error with full context and stack trace
   */
  error(message: string, error?: any, data?: any): void {
    const errorData = {
      ...data,
      error: error instanceof Error ? {
        name: error.name,
        message: error.message,
        stack: error.stack,
      } : error,
    };
    console.error(this.formatMessage('ERROR', message, errorData));
  }

  /**
   * Log debug message (useful for development)
   */
  debug(message: string, data?: any): void {
    if (process.env.LOG_LEVEL === 'DEBUG') {
      console.log(this.formatMessage('DEBUG', message, data));
    }
  }

  /**
   * Log handler entry point
   */
  logRequest(event: any): void {
    this.info('Handler invoked', {
      queryStringParameters: event.queryStringParameters,
      pathParameters: event.pathParameters,
      hasBody: !!event.body,
      bodyLength: event.body?.length,
    });
  }

  /**
   * Log handler response
   */
  logResponse(statusCode: number, body?: any): void {
    this.info('Handler response', {
      statusCode,
      bodyLength: typeof body === 'string' ? body.length : JSON.stringify(body)?.length,
    });
  }

  /**
   * Log DynamoDB operation
   */
  logDbOperation(operation: string, table: string, params?: any): void {
    this.debug(`DynamoDB ${operation}`, { table, params });
  }

  /**
   * Log S3 operation
   */
  logS3Operation(operation: string, bucket: string, key?: string): void {
    this.debug(`S3 ${operation}`, { bucket, key });
  }
}

// Export singleton instance
export const logger = new Logger();

// Export function to create new logger per request (recommended for concurrent handlers)
export function createLogger(event: any, context?: any): Logger {
  const log = new Logger();
  log.init(event, context);
  return log;
}

File: src/python_handlers/logger.py
"""
Structured logger utility for Lambda handlers
Provides consistent logging format with context for debugging
"""

import json
import logging
import traceback
from datetime import datetime
from typing import Any, Dict, Optional


class Logger:
    """Structured logger for Lambda handlers with JSON output"""
    
    def __init__(self):
        self.context: Dict[str, Any] = {}
        self._logger = logging.getLogger()
        self._logger.setLevel(logging.INFO)
    
    def init(self, event: Dict[str, Any], lambda_context: Any = None) -> None:
        """Initialize logger with request context"""
        request_context = event.get('requestContext', {})
        http_context = request_context.get('http', {})
        
        self.context = {
            'requestId': (
                getattr(lambda_context, 'aws_request_id', None) or 
                request_context.get('requestId', 'unknown')
            ),
            'path': http_context.get('path') or event.get('rawPath'),
            'method': http_context.get('method'),
        }
    
    def _format_message(self, level: str, message: str, data: Optional[Dict] = None) -> str:
        """Format message with context as JSON"""
        log_entry = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'level': level,
            'requestId': self.context.get('requestId'),
            'path': self.context.get('path'),
            'method': self.context.get('method'),
            'message': message,
        }
        
        if data:
            log_entry['data'] = data
            
        return json.dumps(log_entry, default=str)
    
    def info(self, message: str, data: Optional[Dict] = None) -> None:
        """Log informational message"""
        print(self._format_message('INFO', message, data))
    
    def warn(self, message: str, data: Optional[Dict] = None) -> None:
        """Log warning message"""
        print(self._format_message('WARN', message, data))
    
    def error(self, message: str, error: Optional[Exception] = None, data: Optional[Dict] = None) -> None:
        """Log error with full context and stack trace"""
        error_data = dict(data) if data else {}
        
        if error:
            error_data['error'] = {
                'type': type(error).__name__,
                'message': str(error),
                'traceback': traceback.format_exc()
            }
        
        print(self._format_message('ERROR', message, error_data))
    
    def debug(self, message: str, data: Optional[Dict] = None) -> None:
        """Log debug message (only when LOG_LEVEL=DEBUG)"""
        import os
        if os.environ.get('LOG_LEVEL') == 'DEBUG':
            print(self._format_message('DEBUG', message, data))
    
    def log_request(self, event: Dict[str, Any]) -> None:
        """Log handler entry point"""
        self.info('Handler invoked', {
            'queryStringParameters': event.get('queryStringParameters'),
            'pathParameters': event.get('pathParameters'),
            'hasBody': bool(event.get('body')),
            'bodyLength': len(event.get('body', '') or ''),
        })
    
    def log_response(self, status_code: int, body: Any = None) -> None:
        """Log handler response"""
        body_str = json.dumps(body) if body else ''
        self.info('Handler response', {
            'statusCode': status_code,
            'bodyLength': len(body_str),
        })


def create_logger(event: Dict[str, Any], context: Any = None) -> Logger:
    """Create and initialize a new logger for the request"""
    logger = Logger()
    logger.init(event, context)
    return logger


